internal/controller/numaflowcontrollerrollout_controller_test.go

```diff
@@ -36,48 +42,102 @@ var _ = Describe("NumaflowControllerRollout Controller", func() {
 
 		typeNamespacedName := types.NamespacedName{
 			Name:      resourceName,
-			Namespace: "default", // TODO(user):Modify as needed
+			Namespace: "default",
 		}
-		numaflowcontrollerrollout := &apiv1.NumaflowControllerRollout{}
+		numaflowControllerRollout := &apiv1.NumaflowControllerRollout{}
 
 		BeforeEach(func() {
 			By("creating the custom resource for the Kind NumaflowControllerRollout")
-			err := k8sClient.Get(ctx, typeNamespacedName, numaflowcontrollerrollout)
+			err := k8sClient.Get(ctx, typeNamespacedName, numaflowControllerRollout)
 			if err != nil && errors.IsNotFound(err) {
 				resource := &apiv1.NumaflowControllerRollout{
 					ObjectMeta: metav1.ObjectMeta{
 						Name:      resourceName,
 						Namespace: "default",
 					},
-					// TODO(user): Specify other spec details if needed.
+					Spec: apiv1.NumaflowControllerRolloutSpec{
+						Controller: apiv1.Controller{Version: "1.2.0"},
+					},
 				}
 				Expect(k8sClient.Create(ctx, resource)).To(Succeed())
 			}
+
+			// Initialize the controller reconciler configuration
+			stateCache := sync.NewLiveStateCache(cfg)
+			numaLogger := logger.GetBaseLogger().WithName("state cache").WithValues("numaflowcontrollerrollouttest")
+			Expect(stateCache.Init(numaLogger)).To(Succeed())
+			controllerReconciler = &NumaflowControllerRolloutReconciler{
+				client:     k8sClient,
+				scheme:     k8sClient.Scheme(),
+				restConfig: cfg,
+				rawConfig:  cfg,
+				stateCache: stateCache,
+				kubectl:    kubernetes.NewKubectl(),
+			}
 		})
 
 		AfterEach(func() {
-			// TODO(user): Cleanup logic after each test, like removing the resource instance.
 			resource := &apiv1.NumaflowControllerRollout{}
 			err := k8sClient.Get(ctx, typeNamespacedName, resource)
 			Expect(err).NotTo(HaveOccurred())
 
 			By("Cleanup the specific resource instance NumaflowControllerRollout")
 			Expect(k8sClient.Delete(ctx, resource)).To(Succeed())
 		})
-		// It("should successfully reconcile the resource", func() {
-		// 	By("Reconciling the created resource")
-		// 	controllerReconciler := &NumaflowControllerRolloutReconciler{
-		// 		client:     k8sClient,
-		// 		scheme:     k8sClient.Scheme(),
-		// 		restConfig: cfg,
-		// 	}
-
-		// 	_, err := controllerReconciler.Reconcile(ctx, sigsReconcile.Request{
-		// 		NamespacedName: typeNamespacedName,
-		// 	})
-		// 	Expect(err).NotTo(HaveOccurred())
-		// 	// TODO(user): Add more specific assertions depending on your controller's reconciliation logic.
-		// 	// Example: If you expect a certain status condition after reconciliation, verify it here.
-		// })
+
+		It("should successfully reconcile the resource", func() {
+			By("Reconciling the created resource")
```

Comment:
you mean `numaflow-controller` right?

Reply:
For checking that `numaflow-controller` is running I have to run the numaflow controller as well with all dependent configs which I think will be a kind of e2e test instead of unit test? what do you think?

Reply:
For checking that `numaflow-controller` is running I have to run the numaflow controller as well with all dependent configs which I think will be a kind of e2e test instead of unit test? what do you think?

Reply:
even as a unit test, we'd like to verify the reconciliation logic is as expected for individual controller, with envtest that provides the mocked cluster, ideally we would want to verify the underlying numaflow controller as well. Then in the e2e test, we are going to verify in a real cluster with broader cross component coverage. Thanks!